development/extensions/tutorial_testing.rst

```diff
@@ -0,0 +1,913 @@
+
+            $this->helper->expects($this->once())
+                ->method('render')
+                ->with('demo_body.html',test_handle_bertie() $name, 200, false)
+                ->willReturn($this->getMock('Symfony\Component\HttpFoundation\Response'));
+
+            $response = $this->controller->handle($name);
+            $this->assertInstanceOf('\Symfony\Component\HttpFoundation\Response', $response);
+        }
+    }
+
+Using mocks
+-----------
+
+In the ``setUp()`` method we create our controller object. However, we do not
+use the actual classes of phpBB which are used by the controller, when opening
+the page. Instead
+`phpunit mocks <https://phpunit.de/manual/current/en/test-doubles.html>`_ are
+injected.
+
+These mocks allow to check how often a method is called, what the arguments are
+and let you specify the return value. This helps us to verify that **our**
+controller code behaves as expected. We do not want to get an error, when phpBB
+itself behaves unexpected.
+
+Testing a method
+----------------
+
+In our controller we added special handling when the provided name is
+``bertie``. So our first test is whether the
+``phpbb\controller\helper::message()`` method is called **once** with the
+``NO_AUTH_SPEAKING`` error message. We also want to make sure that the
+``phpbb\controller\helper::render()`` method is not called in this case.
+
+Then we call the controller's ``handle()`` method. If the methods are invoked
+correctly the test will pass.
+
+.. code-block:: php
+
+        public function test_handle_bertie()
+        {
+            $this->helper->expects($this->once())
+                ->method('message')
+                ->with('NO_AUTH_SPEAKING', array('bertie'), 'NO_AUTH_OPERATION', 403)
+                ->willReturn($this->getMock('Symfony\Component\HttpFoundation\Response'));
+
+            $this->helper->expects($this->never())
+                ->method('render');
+
+            $response = $this->controller->handle('bertie');
+            $this->assertInstanceOf('\Symfony\Component\HttpFoundation\Response', $response);
+        }
+
+.. note::
+
+    Make sure that the name of your testing method starts with ``test``.
+    Otherwise it is not being executed as a test by phpunit.
+
+Data providers
+--------------
+
+For the normal case we use a data provider. A data provider is a method that
+returns an array of arrays. The inner array contains the arguments for the
+``test`` method.
+
+.. code-block:: php
+
+        public function handle_data()
+        {
+            return array(
+                array('foo', true, 'DEMO_GOODBYE'),
+                array('bar', false, 'DEMO_HELLO'),
+            );
+        }
+
+So in our case the test will be called twice, once with the arguments:
+
+* 'foo'
+* true
+* 'DEMO_GOODBYE'
+
+and a second time with:
+
+* 'bar'
+* false
+* 'DEMO_HELLO'
+
+In the test we then tell the ``phpbb\config\config`` mock to return the
```

Comment:
...mock to expect to be called only once, and to return the specified value (which is passed in as an argument) for the 'acme_demo_goodbye' config variable.